### PR TITLE
Safe CMA-ES (GECCO 2024)

### PR DESCRIPTION
**Pull request by kento031**
This is the implementation of Safe CMA-ES [Uchida+, GECCO 2024].
https://arxiv.org/abs/2405.10534

### CODE REVIEW

**Comment by nomuramasahir0** [comment]
@c-bata 
This PR introduces a new method called SafeCMA, which will require `gpytorch` and `torch` (for Gaussian processes) when users choose to use it. I believe this PR won’t affect Optuna since SafeCMA is unrelated to Optuna, but could you please double-check to ensure there aren’t any issues?

File: README.md [hunk 1]
```diff
@@ -463,3 +463,4 @@ Rate Adaptation: Can CMA-ES with Default Population Size Solve Multimodal
 and Noisy Problems?, GECCO, 2023.](https://arxiv.org/abs/2304.03473)
 * [Nomura and Shibata 2024] [M. Nomura, M. Shibata, cmaes : A Simple yet Practical Python Library for CMA-ES, arXiv:2402.01373, 2024.](https://arxiv.org/abs/2402.01373)
 * [Ros and Hansen 2008] [R. Ros, N. Hansen, A Simple Modification in CMA-ES Achieving Linear Time and Space Complexity, PPSN, 2008.](https://hal.inria.fr/inria-00287367/document)
+* [Uchida et al. 2024] [K. Uchida, R. Hamano, M. Nomura, S. Saito, S. Shirakawa, CMA-ES for Safe Optimization, GECCO, 2024.](https://arxiv.org/abs/2405.10534)
```
Comment (nomuramasahir0) [inline README.md:466]: Could you provide a brief description and usage for SafeCMA (like [LRA-CMA-ES](https://github.com/CyberAgentAILab/cmaes?tab=readme-ov-file#learning-rate-adaptation-cma-es-nomura-et-al-2023))?

File: cmaes/_safe_cma.py [hunk 2]
```diff
@@ -0,0 +1,702 @@
...
+
+        seed:
+            A seed number (optional).
+
+        population_size:
+            A population size (optional).
+
+        cov:
+            A covariance matrix (optional).
+    """
```
Comment (nomuramasahir0) [inline cmaes/_safe_cma.py:122]: Please add the paper link like [this](https://github.com/CyberAgentAILab/cmaes/blob/main/cmaes/_catcma.py#L81).

File: README.md (hunk 1 above)
Comment (nomuramasahir0) [reply README.md:466]: It would be clearer to explicitly state that the current implementation assumes no noise.

File: cmaes/_safe_cma.py [hunk 3]
```diff
@@ -0,0 +1,702 @@
...
+        cmu = min(
+            1 - c1 - 1e-8,  # 1e-8 is for large popsize.
+            alpha_cov
+            * (mu_eff - 2 + 1 / mu_eff)
+            / ((n_dim + 2) ** 2 + alpha_cov * mu_eff / 2),
+        )
+        assert c1 <= 1 - cmu, "invalid learning rate for the rank-one update"
+        assert cmu <= 1 - c1, "invalid learning rate for the rank-μ update"
+
+        cm = 1  # (eq. 54)
```
Comment (nomuramasahir0) [inline cmaes/_safe_cma.py:196]: I suggest removing comments like eq. ~ (for Hansen's Tutorial paper) in this PR, as they could be confusing when compared with the equations in the SafeCMA paper.

Comment (nomuramasahir0) [reply cmaes/_safe_cma.py:196]: On the other hand, adding equation numbers for SafeCMA-specific operations would improve clarity.

File: cmaes/_safe_cma.py [hunk 4]
```diff
@@ -0,0 +1,702 @@
...
+
+        self._funhist_term = 10 + math.ceil(30 * n_dim / population_size)
+        self._funhist_values = np.empty(self._funhist_term * 2)
+
+    def _compute_lipschitz_constant(self) -> np.ndarray:
+
+        likelihood = gpytorch.likelihoods.GaussianLikelihood(
+            noise_constraint=gpytorch.constraints.GreaterThan(0)
+        )
+        likelihood.noise = 0
```
Comment (nomuramasahir0) [inline cmaes/_safe_cma.py:277]: I'm curious about the validity of this setting because, as mentioned in this thread:
> Some small value, but don't make it too small or numerical performance will suffer. I recommend 1e-4.
https://github.com/cornellius-gp/gpytorch/issues/1196#issuecomment-651193968

This implies there's room for improvement in this setting, but I'm not sure for now, so let's see how it goes.

File: cmaes/_safe_cma.py [hunk 5]
```diff
@@ -0,0 +1,702 @@
...
+            )
+            mean = likelihood(model(X)).mean
+            dxdmean = torch.autograd.grad(mean.sum(), X)[0]
+
+            grad_norm = torch.sqrt(torch.sum(dxdmean * dxdmean, dim=1))
+
+            if out_scalar:
+                grad_norm = grad_norm.mean().to(torch.float64)
+
+            return -grad_norm
```
Comment (nomuramasahir0) [inline cmaes/_safe_cma.py:310]: `df` is defined as `# function that returns the norm of gradient`, but it currently returns a negative value, which seems odd. 
I understand this is intended for `scipy.optimize.minimize`, but it would be better to align it with the description.

File: cmaes/_safe_cma.py [hunk 6]
```diff
@@ -0,0 +1,702 @@
...
+            # safety function
+            def safe_function(x):
+                return x[0]
+
+            # safe seeds
+            safe_seeds_num = 10
+            safe_seeds = (np.random.rand(safe_seeds_num, dim) * 2 - 1) * 5
+            safe_seeds[:, 0] = - np.abs(safe_seeds[:, 0])
+
+            # evaluation of safe seeds (with multiple safety functions)
```
Comment (nomuramasahir0) [inline cmaes/_safe_cma.py:47]: > with multiple safety functions

with a single safety function?

File: examples/safecma.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    # safety function
+    def safe_function(x):
+        return x[0]
+
+    # safe seeds
+    safe_seeds_num = 10
+    safe_seeds = (np.random.rand(safe_seeds_num, dim) * 2 - 1) * 5
+    safe_seeds[:, 0] = -np.abs(safe_seeds[:, 0])
+
+    # evaluation of safe seeds (with multiple safety functions)
```
Comment (nomuramasahir0) [inline examples/safecma.py:27]: > with multiple safety functions
with a single safety function?

**Comment by nomuramasahir0** [comment]
@kento031 
Thanks for the PR! I've left some comments. Could you take a look?

File: cmaes/__init__.py [hunk 8]
```diff
@@ -6,4 +6,9 @@
 from ._dxnesic import DXNESIC  # NOQA
 from ._catcma import CatCMA  # NOQA
 
+try:
+    from ._safe_cma import SafeCMA  # NOQA
+except ImportError:
+    pass  # Implementation of Safe CMA-ES requires scipy, gpytorch, and torch
+
```
Comment (c-bata) [inline cmaes/__init__.py:13]: How about simply removing `SafeCMA` and renaming `cmaes/_safe_cma.py` to `cmaes/safe_cma.py`?

```suggestion
```

Comment (nomuramasahir0) [reply cmaes/__init__.py:13]: Thanks for the great suggestion! It looks good to me.

@kento031 
Could you make this modification if it’s okay?

**Comment by kento031** [comment]
@nomuramasahir0 @c-bata 
Thank you for your kind comments. I will address them.

**Comment by kento031** [comment]
Sorry for being so late in addressing your comments. 
I modified my implementation according to your comments.

I checked the code using `act` (https://github.com/nektos/act), and there remains an error when executing `act -j lint` (and when executing `act pull_request`). This error was caused at `cmaes/_cma.py:480` by running `mypy cmaes`. I leave this error as is for now because I did not change `cmaes/_cma.py`.

@nomuramasahir0 We previously discussed about whether `_safe_cma.py` sould be contained in `cmaes/__init__.py`. However, at that time, I misunderstood how the workflow runs. Finally, I modified my implementation according to @c-bata 's comment.

File: cmaes/_safe_cma.py (hunk 4 above)
Comment (kento031) [reply cmaes/_safe_cma.py]: This setting was used in some previous studies, including the followings.
https://link.springer.com/chapter/10.1007/978-3-030-58112-1_13
https://dl.acm.org/doi/10.1007/978-3-031-30229-9_51

In addition, I run the safe CMA-ES with `likelihood.noise = 1e-4` on some benchmark functions. As a result, it occurred a few constraint violations that was not observed with `likelihood.noise = 0`. Therefore, I did not change it.

File: cmaes/_safe_cma.py (hunk 5 above)
Comment (kento031) [reply cmaes/_safe_cma.py]: I changed the explanation of the function as `# function that returns the negative norm of gradient`.

**Comment by ha-mano** [comment]
@kento031 
I have fixed a mypy error that was causing in the workflow tests. The fix is included in [this pull request](https://github.com/CyberAgentAILab/cmaes/pull/195)

**Comment by kento031** [comment]
@ha-mano 
Thank you for fixing the error. I have merged your fix.

File: requirements-dev.txt [hunk 9]
```diff
@@ -1,5 +1,7 @@
 # install_requires
 numpy>=1.20.0
+torch
+gpytorch
```
[the remark this answers is not in the document]
Comment (kento031) [reply requirements-dev.txt:6]: Thank you for your suggestion. I have merged it.

**Review by c-bata — APPROVED** [review]
LGTM. Thank you for your contribution :100:

**Comment by kento031** [comment]
Thank you for all your comments!

In the CI check, the `lint` job failed due to `tools/cmaes_visualizer.py` and `tools/ws_cmaes_visualizer.py`. However, I did not change these files.

Should I modify these codes to resolve the errors?

**Comment by c-bata** [comment]
@kento031 No, you don't need to fix that errors in this PR 👍
